examples/dreambooth/train_dreambooth_lora.py

```diff
@@ -1051,12 +1047,12 @@ def compute_text_embeddings(prompt):
 
     # Prepare everything with our `accelerator`.
     if args.train_text_encoder:
-        unet_lora_layers, text_encoder_lora_layers, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
-            unet_lora_layers, text_encoder_lora_layers, optimizer, train_dataloader, lr_scheduler
+        unet, text_encoder, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
+            unet, text_encoder, optimizer, train_dataloader, lr_scheduler
         )
     else:
-        unet_lora_layers, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
-            unet_lora_layers, optimizer, train_dataloader, lr_scheduler
+        unet, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
+            unet, optimizer, train_dataloader, lr_scheduler
         )
```

Comment:
It's not clear to me that we should be passing a subset of the model parameters to accelerate.prepare. The docs say its used for distributed training and mixed precision which afaik should be applied to the whole model

Reply:
With your current changes, it makes sense to me, especially with the changes you proposed for serialization and loading a couple blocks earlier. 

I think the reason why we were passing the subset of model layers is because of how we handled them internally for serialization and loading.

Reply:
Hmm this needs a bit more testing before being merged. Note that pure fp16 != torch.autocast. 
Accelerate.prepare puts the model into autocast see [here](https://github.com/huggingface/accelerate/blob/47b9ac10710dd51e6e4e1304223d6dc543c47a54/src/accelerate/accelerator.py#L1306).

The reason we didn't pass the whole model to begin with was so that we can use 99% of the model in pure fp16 and don't have to autocast it (fp16 autocast is different as e.g. it puts the layer norm in fp32). 

I agree that it's cleaner to put the whole model into prepare, but we need to check that we don't significantly increase our memory consumption here. Could you run a quick test here @williamberman ?

Reply:
I think we have to make this change regardless of effects on the memory as we're misusing the accelerate api by passing the dummy module to it. 

I.e. in the case of distributed training accelerate expects to receive a whole model that it then handles wrapping in ddp. If it received the dummy model wrapping the lora weights, it leaves the whole model on one gpu and then just puts the dummy model in ddp, and I don't think that makes sense? 

Similarly in the linked code where it wraps the prepared model's forward method in amp, the dummy model doesn't have a forward method so really if we were configuring accelerate for amp, I don't think we'd be doing any mixed precision with the lora weights?

These are differences in memory usage for a forward/backward of the unet with a single item.

```py
import torch
from diffusers import UNet2DConditionModel

unet = UNet2DConditionModel.from_pretrained('runwayml/stable-diffusion-v1-5', torch_dtype=torch.float32, subfolder="unet")
unet.to('cuda')

torch.cuda.reset_max_memory_allocated()
sample = torch.rand((1, 4, 64, 64), dtype=torch.float32, device='cuda')
timestep = torch.tensor(1, device='cuda')
encoder_hidden_states = torch.rand((1, 77, 768), dtype=torch.float32, device='cuda')
unet(sample=sample, timestep=timestep, encoder_hidden_states=encoder_hidden_states).sample.sum().backward()
print("float 32")
print(torch.cuda.max_memory_allocated('cuda') / 10**9)

torch.cuda.reset_max_memory_allocated()
sample = torch.rand((1, 4, 64, 64), dtype=torch.float32, device='cuda')
timestep = torch.tensor(1, device='cuda')
encoder_hidden_states = torch.rand((1, 77, 768), dtype=torch.float32, device='cuda')
with torch.cuda.amp.autocast():
    unet(sample=sample, timestep=timestep, encoder_hidden_states=encoder_hidden_states).sample.sum().backward()
print("float 32 + amp")
print(torch.cuda.max_memory_allocated('cuda') / 10**9)


 
unet = UNet2DConditionModel.from_pretrained('runwayml/stable-diffusion-v1-5', torch_dtype=torch.float16, subfolder="unet")
unet.to('cuda')

torch.cuda.reset_max_memory_allocated()
sample = torch.rand((1, 4, 64, 64), dtype=torch.float16, device='cuda')
timestep = torch.tensor(1, device='cuda')
encoder_hidden_states = torch.rand((1, 77, 768), dtype=torch.float16, device='cuda')
unet(sample=sample, timestep=timestep, encoder_hidden_states=encoder_hidden_states).sample.sum().backward()
print("float 16")
print(torch.cuda.max_memory_allocated('cuda') / 10**9)

torch.cuda.reset_max_memory_allocated()
sample = torch.rand((1, 4, 64, 64), dtype=torch.float16, device='cuda')
timestep = torch.tensor(1, device='cuda')
encoder_hidden_states = torch.rand((1, 77, 768), dtype=torch.float16, device='cuda')
with torch.cuda.amp.autocast():
    unet(sample=sample, timestep=timestep, encoder_hidden_states=encoder_hidden_states).sample.sum().backward()
print("float 16 + amp")
print(torch.cuda.max_memory_allocated('cuda') / 10**9)
```

float 32
7.528367616
float 32 + amp
10.071896576
float 16
3.856464896
float 16 + amp
4.879464448

Reply:
I don't understand this: https://github.com/huggingface/diffusers/pull/3778/files#r1248373203 - we want to measure memory usage for LoRA training for specific batch sizes

Reply:
Also why is `unet_lora_layers` a dummy model? It's PyTorch weights

Reply:
chatted with patrick about this over slack, confirmed that as is in main, the script doesn't correctly let accelerate wrap the unet and text encoder so amp/ddp aren't actually being done i.e. we can see weights diverge on the different GPUs

also confirmed that memory increase when using amp is small (even compared to before where amp wasn't actually being done)